File: typescript/src/governanceWebhooks.ts
/**
 * Governance and enforcement webhook types — wire shapes for
 * `v1-governance-webhooks` and `v1-enforcement-webhooks`.
 *
 * Covers subscription management, signed delivery callbacks, and
 * delivery receipt records. Use `verifyWebhookSignature` in your
 * receiver to authenticate payloads.
 */

export type GovernanceWebhookEvent =
  | "enforcement.blocked"
  | "policy.violation"
  | "access_review.completed"
  | "mfa.enrollment_required"
  | "contract.activated"
  | "replay.drift_detected";

export type EnforcementWebhookEvent =
  | "enforcement.blocked"
  | "enforcement.pending_approval"
  | "enforcement.approved"
  | "enforcement.expired";

export type WebhookDeliveryStatus = "pending" | "delivered" | "failed";

export interface WebhookSubscription {
  id: string;
  org_id: string;
  url: string;
  events: string[];
  enabled: boolean;
  description: string | null;
  created_at: string;
}

export interface WebhookDelivery {
  id: string;
  subscription_id: string;
  event_type: string;
  payload: Record<string, unknown>;
  status: WebhookDeliveryStatus;
  response_status: number | null;
  response_body: string | null;
  attempted_at: string;
}

export interface CreateWebhookSubscriptionRequest {
  url: string;
  events: string[];
  enabled?: boolean;
  description?: string;
}

export interface ListWebhookSubscriptionsResponse {
  subscriptions: WebhookSubscription[];
  /** For governance webhooks, the server returns supported event names. */
  valid_events?: string[];
}

export interface ListWebhookDeliveriesResponse {
  deliveries: WebhookDelivery[];
}

/**
 * Signed webhook payload delivered to subscriber endpoints.
 *
 * AtlaSent sets two headers on every delivery:
 * - `X-AtlaSent-Signature: sha256=<hex>` — HMAC-SHA256 of the raw body
 * - `X-AtlaSent-Event: <event_type>`
 *
 * Verify with `verifyWebhookSignature` before processing.
 */
export interface WebhookPayload<T = Record<string, unknown>> {
  id: string;
  org_id: string;
  event_type: string;
  source_id: string | null;
  data: T;
  created_at: string;
}

/**
 * Verify a webhook delivery signature using the subscription secret.
 *
 * Performs a constant-time HMAC-SHA256 comparison so timing attacks
 * cannot reveal secret length. Requires `globalThis.crypto.subtle`
 * (Node 20+, all modern browsers, Deno, Cloudflare Workers).
 *
 * @param payload   Raw UTF-8 request body string.
 * @param signature Value of the `X-AtlaSent-Signature` header.
 * @param secret    Webhook subscription secret from the AtlaSent console.
 * @returns `true` when the signature is valid.
 *
 * @example
 * ```ts
 * import { verifyWebhookSignature } from "@atlasent/sdk";
 *
 * app.post("/webhook", async (req, res) => {
 *   const ok = await verifyWebhookSignature(
 *     req.rawBody,
 *     req.headers["x-atlasent-signature"],
 *     process.env.WEBHOOK_SECRET!,
 *   );
 *   if (!ok) return res.status(401).send("invalid signature");
 *   // process req.body …
 * });
 * ```
 */
export async function verifyWebhookSignature(
  payload: string,
  signature: string,
  secret: string,
): Promise<boolean> {
  const prefix = "sha256=";
  if (!signature.startsWith(prefix)) return false;
  const receivedHex = signature.slice(prefix.length);

  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sigBuffer = await crypto.subtle.sign("HMAC", key, encoder.encode(payload));
  const expectedHex = Array.from(new Uint8Array(sigBuffer))
    .map(b => b.toString(16).padStart(2, "0"))
    .join("");

  if (receivedHex.length !== expectedHex.length) return false;
  let diff = 0;
  for (let i = 0; i < expectedHex.length; i++) {
    diff |= receivedHex.charCodeAt(i) ^ expectedHex.charCodeAt(i);
  }
  return diff === 0;
}

File: python/atlasent/governance_webhooks.py
"""Governance and enforcement webhook types and signature verification.

Parity with ``typescript/src/governanceWebhooks.ts``.

Covers two webhook surfaces:

- ``/v1/governance-webhooks`` — governance-layer events (policy
  violations, runtime anomalies, HITL decisions, audit chain entries,
  billing state changes).
- ``/v1/enforcement-webhooks`` — real-time enforcement decision events
  (blocked actions, pending approvals, approvals/rejections).

Both surfaces use the same HMAC-SHA256 signing scheme.  Use
:func:`verify_webhook_signature` to authenticate incoming deliveries
before processing their payloads.
"""

from __future__ import annotations

import hashlib
import hmac
from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Event type literals
# ---------------------------------------------------------------------------

GovernanceWebhookEvent = Literal[
    "policy.violation",
    "policy.updated",
    "runtime.anomaly",
    "runtime.chain_broken",
    "hitl.decision",
    "hitl.timeout",
    "audit.chain_entry",
    "billing.grace_period_started",
    "billing.access_restricted",
]
"""Event types deliverable via ``/v1/governance-webhooks``."""

EnforcementWebhookEvent = Literal[
    "enforcement.blocked",
    "enforcement.pending_approval",
    "enforcement.approved",
    "enforcement.rejected",
    "enforcement.error",
]
"""Event types deliverable via ``/v1/enforcement-webhooks``."""

WebhookDeliveryStatus = Literal["pending", "delivered", "failed", "retrying"]

# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------


class WebhookSubscription(BaseModel):
    """A registered webhook subscription returned by the API."""

    id: str
    org_id: str
    url: str
    events: list[str]
    enabled: bool = True
    description: str | None = None
    created_at: str = ""

    model_config = {"extra": "allow"}


class WebhookDelivery(BaseModel):
    """One delivery attempt record for a webhook subscription."""

    id: str
    subscription_id: str
    event_type: str
    payload: dict[str, Any] = Field(default_factory=dict)
    status: WebhookDeliveryStatus
    response_status: int | None = None
    response_body: str | None = None
    attempt_count: int = 1
    created_at: str = ""
    delivered_at: str | None = None

    model_config = {"extra": "allow"}


T = TypeVar("T")


class WebhookPayload(BaseModel, Generic[T]):
    """Envelope for every webhook delivery body.

    The ``data`` field contains the event-specific payload.  Its shape
    varies by ``event_type``; callers should validate it against the
    appropriate model for the event they are handling.
    """

    id: str
    org_id: str
    event_type: str
    source_id: str | None = None
    data: Any = None
    created_at: str = ""

    model_config = {"extra": "allow"}


# ---------------------------------------------------------------------------
# Signature verification
# ---------------------------------------------------------------------------


def verify_webhook_signature(
    payload: bytes | str,
    signature: str,
    secret: str,
) -> bool:
    """Verify the HMAC-SHA256 signature on an incoming webhook delivery.

    The ``X-AtlaSent-Signature`` header value has the form
    ``sha256=<hex-digest>``.  This function computes the expected
    HMAC-SHA256 over the **raw request body bytes** using *secret* and
    compares with *signature* using a constant-time comparison to
    prevent timing attacks.

    Parameters
    ----------
    payload:
        Raw request body.  Pass the bytes *before* any JSON parsing —
        parsing may change whitespace and break the signature.
    signature:
        Value of the ``X-AtlaSent-Signature`` header, e.g.
        ``"sha256=abc123..."``.
    secret:
        The ``signing_secret`` returned when the subscription was
        created.  Store it as an environment variable and never log it.

    Returns
    -------
    bool
        ``True`` when the signature is valid, ``False`` otherwise.
        **Always reject the payload when this returns** ``False``.

    Examples
    --------
    FastAPI::

        from fastapi import FastAPI, Request, HTTPException
        from atlasent import verify_webhook_signature

        @app.post("/webhooks/atlasent")
        async def handle(request: Request):
            raw = await request.body()
            sig = request.headers.get("x-atlasent-signature", "")
            if not verify_webhook_signature(raw, sig, WEBHOOK_SECRET):
                raise HTTPException(401, "Invalid signature")
            ...
    """
    body_bytes: bytes = (
        payload if isinstance(payload, bytes) else payload.encode("utf-8")
    )
    secret_bytes: bytes = (
        secret if isinstance(secret, bytes) else secret.encode("utf-8")
    )

    expected_hex = hmac.new(secret_bytes, body_bytes, hashlib.sha256).hexdigest()
    expected = f"sha256={expected_hex}"

    # Strip the "sha256=" prefix from the received value for comparison.
    # hmac.compare_digest requires both operands to be the same type.
    return hmac.compare_digest(
        expected.encode("utf-8"),
        signature.encode("utf-8"),
    )
